### PR TITLE
Support for inheriting public api annotations (#1773)

### PR DESCRIPTION
**Pull request by tobiasstadler**
## What does this PR do?
Fixes elastic/apm-agent-java#1773

## Checklist
- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] Added an API method or config option? Document in which version this will be introduced
  - [x] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [ ] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)

### CODE REVIEW

**Review by SylvainJuge — APPROVED** [review]
LGTM, approval of PR to enable test on CI

File: apm-agent-plugins/apm-api-plugin/src/test/java/co/elastic/apm/api/AnnotationInheritanceTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+class AnnotationInheritanceTest {
+
+    protected static ElasticApmTracer tracer;
+    protected static MockReporter reporter;
+
+    @BeforeAll
+    static void beforeAll() {
+        MockTracer.MockInstrumentationSetup mockInstrumentationSetup = MockTracer.createMockInstrumentationSetup();
+        tracer = mockInstrumentationSetup.getTracer();
+        when(tracer.getConfig(CoreConfiguration.class).isEnablePublicapiAnnotationInheritance()).thenReturn(true);
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-api-plugin/src/test/java/co/elastic/apm/api/AnnotationInheritanceTest.java:54]: To complete the tests, also test that:
- inheritance doesn't work by default (`enable_publicapi_annotation_inheritance == false`)
- `ClassWithAnnotations`'s methods are traced with either configuration of `enable_publicapi_annotation_inheritance`

File: apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java [hunk 2]
```diff
@@ -788,6 +799,10 @@ public CloudProvider getCloudProvider() {
         return cloudProvider.get();
     }
 
+    public boolean isEnablePublicapiAnnotationInheritance() {
+        return enablePublicapiAnnotationInheritance.get();
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java:803]:
```suggestion
    public boolean isEnablePublicApiAnnotationInheritance() {
        return enablePublicApiAnnotationInheritance.get();
```

**Review by eyalkoren — COMMENTED** [review]
Looks great, well done! 👏 
A few minor suggestions.
For the config option name and description - you would need to verify that the field and method names are properly set, and you will have to run `ConfigurationExporterTest` so that `configuration.asciidoc` gets updated.
Thanks a lot, this is a very useful extension to our API!

**Comment by tobiasstadler** [comment]
@eyalkoren Thank you for updating configuration.asciidoc!

**Review by eyalkoren — APPROVED** [review]
Awesome work, thanks @tobiasstadler ! ❤️ 

I just did a few small cosmetic changes in the tests and removed the method matching check on super-annotated-methods being in classes within the configured application packages as it is not required.

**Comment by tobiasstadler** [comment]
Thank You!

**Pull request merged**